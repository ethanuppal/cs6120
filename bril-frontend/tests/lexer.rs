use bril_frontend::{lexer::Token, loc::Loc};
use insta::assert_debug_snapshot;
use logos::Logos;

#[allow(dead_code)] // since Debug
#[derive(Debug)]
struct LexerSnapshot<'a> {
    code: &'a str,
    tokens: Vec<Loc<Token<'a>>>,
}

#[allow(dead_code)] // since Debug
#[derive(Debug)]
struct LexerErrorSnapshot<'a> {
    code: &'a str,
    tokens: Vec<Loc<Token<'a>>>,
    failure: Loc<&'a str>,
    leftover: &'a str,
}

macro_rules! lexer_snapshot {
    ($name:ident, $code:expr) => {
        #[test]
        fn $name() {
            let code = $code;
            let mut lexer = Token::lexer(code);
            let mut tokens = vec![];
            while let Some(next) = lexer.next() {
                if let Ok(token) = next {
                    tokens.push(Loc::new(token, lexer.span()));
                } else {
                    panic!("Failed to lex. Leftover: {}", lexer.remainder());
                }
            }

            assert_debug_snapshot!(LexerSnapshot { code, tokens });
        }
    };
}

macro_rules! lexer_error {
    ($name:ident, $code:expr) => {
        #[test]
        fn $name() {
            let code = $code;
            let mut lexer = Token::lexer(code);
            let mut tokens = vec![];
            while let Some(next) = lexer.next() {
                if let Ok(token) = next {
                    tokens.push(Loc::new(token, lexer.span()));
                } else {
                    break;
                }
            }

            if lexer.remainder().is_empty() {
                panic!("Lexing was unfortunately successful: {:?}", tokens);
            }

            assert_debug_snapshot!(LexerErrorSnapshot {
                code,
                tokens,
                failure: Loc::new(lexer.slice(), lexer.span()),
                leftover: lexer.remainder()
            });
        }
    };
}

lexer_snapshot! {
    all_tokens_work,
    r#"import from as @main .foo foo "path/to/lol" {}:<>;=,() 5 5.0 'a'"#
}

lexer_snapshot! {
    add_bril_lexes,
    include_str!("../bril-programs/add.bril")
}

lexer_snapshot! {
    import_bril_lexes,
    include_str!("../bril-programs/import.bril")
}

lexer_error! {
    invalid_characters,
    "$main"
}
