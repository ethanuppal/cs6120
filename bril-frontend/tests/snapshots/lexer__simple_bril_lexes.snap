---
source: bril-frontend/tests/lexer.rs
expression: "LexerSnapshot { code, tokens }"
---
LexerSnapshot {
    code: "@main() {\n.test:\n  a: int = const 0;\n}\n",
    tokens: [
        Loc {
            inner: FunctionName(
                "@main",
            ),
            span: 0..5,
        },
        Loc {
            inner: LeftPar,
            span: 5..6,
        },
        Loc {
            inner: RightPar,
            span: 6..7,
        },
        Loc {
            inner: LeftBrace,
            span: 8..9,
        },
        Loc {
            inner: Label(
                ".test",
            ),
            span: 10..15,
        },
        Loc {
            inner: Colon,
            span: 15..16,
        },
        Loc {
            inner: Identifier(
                "a",
            ),
            span: 19..20,
        },
        Loc {
            inner: Colon,
            span: 20..21,
        },
        Loc {
            inner: Identifier(
                "int",
            ),
            span: 22..25,
        },
        Loc {
            inner: Equals,
            span: 26..27,
        },
        Loc {
            inner: Identifier(
                "const",
            ),
            span: 28..33,
        },
        Loc {
            inner: Integer(
                0,
            ),
            span: 34..35,
        },
        Loc {
            inner: Semi,
            span: 35..36,
        },
        Loc {
            inner: RightBrace,
            span: 37..38,
        },
    ],
}
